File: ocsgw/src/main/java/com/telenordigital/ocsgw/OcsServer.java
package com.telenordigital.ocsgw;

import com.telenordigital.ocsgw.data.DataSource;
import com.telenordigital.ocsgw.data.DataSourceType;
import com.telenordigital.ocsgw.data.grpc.GrpcDataSource;
import com.telenordigital.ocsgw.data.local.LocalDataSource;
import com.telenordigital.ocsgw.utils.AppConfig;
import com.telenordigital.ostelco.diameter.CreditControlContext;
import com.telenordigital.ostelco.diameter.model.ReAuthRequestType;
import org.jdiameter.api.*;
import org.jdiameter.api.auth.ServerAuthSession;
import org.jdiameter.api.auth.events.ReAuthRequest;
import org.jdiameter.api.cca.ServerCCASession;
import org.jdiameter.api.cca.events.JCreditControlRequest;
import org.jdiameter.client.api.ISessionFactory;
import org.jdiameter.client.impl.parser.MessageImpl;
import org.jdiameter.common.impl.app.auth.AuthSessionFactoryImpl;
import org.jdiameter.common.impl.app.auth.ReAuthRequestImpl;
import org.jdiameter.server.impl.app.cca.ServerCCASessionImpl;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.function.Consumer;


public class OcsServer {

    private static final Logger LOG = LoggerFactory.getLogger(OcsApplication.class);
    private static final OcsServer INSTANCE = new OcsServer();
    private Stack stack;
    private DataSource source;

    public static OcsServer getInstance() {
        return INSTANCE;
    }

    private OcsServer() {
    }

    public synchronized void handleRequest(ServerCCASession session, JCreditControlRequest request) {

        final CreditControlContext ccrContext = new CreditControlContext(session.getSessionId(), request);
        source.handleRequest(ccrContext);
    }

    public Stack getStack() {
        return stack;
    }

    //https://tools.ietf.org/html/rfc4006#page-30
    //https://tools.ietf.org/html/rfc3588#page-101
    public void sendReAuthRequest(final String sessionId, final String originHost, final String originRealm, final String destinationHost, final String destinationRealm) {
        try {
            ServerCCASessionImpl ccaSession = stack.getSession(sessionId, ServerCCASessionImpl.class);
            if (ccaSession != null && ccaSession.isValid()) {
                for (Session session : ccaSession.getSessions() ) {
                    Request request = session.createRequest(258, ApplicationId.createByAuthAppId(4L), destinationRealm, destinationHost);
                    AvpSet avps = request.getAvps();
                    avps.addAvp(Avp.RE_AUTH_REQUEST_TYPE, ReAuthRequestType.AUTHORIZE_ONLY.getValue());
                    avps.addAvp(Avp.ORIGIN_HOST, originHost, true, false, true);
                    avps.addAvp(Avp.ORIGIN_REALM, originRealm, true, false, true);
                    ReAuthRequest reAuthRequest = new ReAuthRequestImpl(request);
                    ccaSession.sendReAuthRequest(reAuthRequest);
                }
            } else {
                LOG.info("No session with ID {}", sessionId);
            }
        } catch (InternalException | IllegalDiameterStateException | RouteException | OverloadException e) {
            LOG.warn("Failed to send Re-Auth Request");
        }
    }

    public void init(Stack stack, AppConfig appConfig) {
        this.stack = stack;

        switch (appConfig.getDataStoreType()) {
            case DataSourceType.GRPC:
                LOG.info("Using GrpcDataSource");
                source = new GrpcDataSource(appConfig.getGrpcServer(), appConfig.encryptGrpc());
                break;
            case DataSourceType.LOCAL:
                LOG.info("Using LocalDataSource");
                source = new LocalDataSource();
                break;
            default:
                LOG.warn("Unknow DataStoreType {}", appConfig.getDataStoreType());
                source = new LocalDataSource();
                break;
        }
        source.init();
    }
}

File: ocsgw/src/main/kotlin/OcsServer.kt
import org.jdiameter.api.ApplicationId
import org.jdiameter.api.Avp
import org.jdiameter.api.IllegalDiameterStateException
import org.jdiameter.api.InternalException
import org.jdiameter.api.OverloadException
import org.jdiameter.api.RouteException
import org.jdiameter.api.Stack
import org.jdiameter.api.cca.ServerCCASession
import org.jdiameter.api.cca.events.JCreditControlRequest
import org.jdiameter.common.impl.app.auth.ReAuthRequestImpl
import org.jdiameter.server.impl.app.cca.ServerCCASessionImpl
import org.ostelco.diameter.CreditControlContext
import org.ostelco.diameter.getLogger
import org.ostelco.diameter.model.ReAuthRequestType
import org.ostelco.diameter.model.SessionContext
import org.ostelco.ocsgw.datasource.DataSource
import org.ostelco.ocsgw.datasource.DataSourceType.Local
import org.ostelco.ocsgw.datasource.DataSourceType.Proxy
import org.ostelco.ocsgw.datasource.DataSourceType.PubSub
import org.ostelco.ocsgw.datasource.DataSourceType.gRPC
import org.ostelco.ocsgw.datasource.SecondaryDataSourceType
import org.ostelco.ocsgw.datasource.local.LocalDataSource
import org.ostelco.ocsgw.datasource.protobuf.GrpcDataSource
import org.ostelco.ocsgw.datasource.protobuf.ProtobufDataSource
import org.ostelco.ocsgw.datasource.protobuf.PubSubDataSource
import org.ostelco.ocsgw.datasource.proxy.ProxyDataSource
import org.ostelco.ocsgw.utils.AppConfig


object OcsServer {

    private val logger by getLogger()

    var stack: Stack? = null
        private set

    private var source: DataSource? = null
    private var localPeerFQDN: String? = null
    private var localPeerRealm: String? = null
    private var defaultRequestedServiceUnit: Long = 0L

    internal fun handleRequest(session: ServerCCASession, request: JCreditControlRequest) {

        val peerFqdn = localPeerFQDN
        if (peerFqdn == null) {
            logger.error("Failed to create CreditControlContext, local peer fqdn is null")
            return
        }

        val peerRealm = localPeerRealm
        if (peerRealm == null) {
            logger.error("Failed to create CreditControlContext, local peer realm is null")
            return
        }

        try {
            val ccrContext = CreditControlContext(
                    session.sessionId,
                    request,
                    peerFqdn,
                    peerRealm
            )
            setDefaultRequestedServiceUnit(ccrContext)
            source?.handleRequest(ccrContext) ?: logger.error("Received request before initialising stack")
        } catch (e: Exception) {
            logger.error("Failed to create CreditControlContext", e)
        }
    }

    // In the case where the Diameter client does not set the Requested-Service-Unit AVP
    // in the Multiple-Service-Credit-Control we need to set a default value.
    private fun setDefaultRequestedServiceUnit(ccrContext: CreditControlContext) {
        ccrContext.creditControlRequest.multipleServiceCreditControls.forEach { mscc ->
                if ( mscc.requested.size == 1 ) {
                    if ( mscc.requested.get(0).total <= 0) {
                        mscc.requested.get(0).total = defaultRequestedServiceUnit
                    }
                }
            }
    }

    //https://tools.ietf.org/html/rfc4006#page-30
    //https://tools.ietf.org/html/rfc3588#page-101
    fun sendReAuthRequest(sessionContext: SessionContext?) {
        try {
            val ccaSession = stack?.getSession(sessionContext?.sessionId, ServerCCASessionImpl::class.java)
            if (ccaSession != null && ccaSession.isValid) {
                for (session in ccaSession.sessions) {
                    if (session.isValid) {
                        val request = session.createRequest(258,
                                ApplicationId.createByAuthAppId(4L),
                                sessionContext?.originRealm,
                                sessionContext?.originHost
                        )
                        request.isProxiable = true
                        val avps = request.avps
                        avps.addAvp(Avp.RE_AUTH_REQUEST_TYPE, ReAuthRequestType.AUTHORIZE_ONLY.ordinal, true, false)
                        val reAuthRequest = ReAuthRequestImpl(request)
                        ccaSession.sendReAuthRequest(reAuthRequest)
                        logger.debug("Sent RAR [{}]", sessionContext?.sessionId)
                    } else {
                        logger.info("Invalid session")
                    }
                }
            } else {
                logger.info("No session with ID {}", sessionContext?.sessionId)
            }
        } catch (e: InternalException) {
            logger.warn("Failed to send Re-Auth Request [{}]", sessionContext?.sessionId, e)
        } catch (e: IllegalDiameterStateException) {
            logger.warn("Failed to send Re-Auth Request [{}]", sessionContext?.sessionId, e)
        } catch (e: RouteException) {
            logger.warn("Failed to send Re-Auth Request [{}]", sessionContext?.sessionId, e)
        } catch (e: OverloadException) {
            logger.warn("Failed to send Re-Auth Request [{}]", sessionContext?.sessionId, e)
        }

    }

    fun init(stack: Stack, appConfig: AppConfig) {
        OcsServer.stack = stack
        localPeerFQDN = stack.metaData.localPeer.uri.fqdn
        localPeerRealm = stack.metaData.localPeer.realmName

        defaultRequestedServiceUnit = appConfig.defaultRequestedServiceUnit

        val protobufDataSource = ProtobufDataSource()

        source = when (appConfig.dataStoreType) {
            Proxy -> {
                logger.info("Using ProxyDataSource")
                val secondary = when (appConfig.secondaryDataStoreType) {
                    SecondaryDataSourceType.PubSub -> getPubSubDataSource(protobufDataSource, appConfig)
                    SecondaryDataSourceType.gRPC -> getGrpcDataSource(protobufDataSource, appConfig)
                }
                secondary.init()
                ProxyDataSource(secondary)
            }
            Local -> {
                logger.info("Using LocalDataSource")
                LocalDataSource()
            }
            PubSub -> {
                logger.info("Using PubSubDataSource")
                getPubSubDataSource(protobufDataSource, appConfig)
            }
            gRPC -> {
                logger.info("Using GrpcDataSource")
                getGrpcDataSource(protobufDataSource, appConfig)
            }
        }
        source?.init()
    }

    private fun getGrpcDataSource(
            protobufDataSource: ProtobufDataSource,
            appConfig: AppConfig): GrpcDataSource =
            GrpcDataSource(
                    protobufDataSource,
                    appConfig.grpcServer)

    private fun getPubSubDataSource(
            protobufDataSource: ProtobufDataSource,
            appConfig: AppConfig): PubSubDataSource =
            PubSubDataSource(protobufDataSource,
                    appConfig.pubSubProjectId,
                    appConfig.pubSubTopicIdForCcr,
                    appConfig.pubSubTopicIdForCca,
                    appConfig.pubSubSubscriptionIdForCca,
                    appConfig.pubSubSubscriptionIdForActivate)
}